pandora-server-directory/src/account/accountSecure.ts

```diff
@@ -294,20 +353,108 @@ function GenerateSimpleToken(): string {
 }
 
-const TOKEN_TYPES: Record<AccountTokenReason, 'secure' | 'simple'> = {
-	[AccountTokenReason.ACTIVATION]: 'simple',
-	[AccountTokenReason.PASSWORD_RESET]: 'simple',
-	[AccountTokenReason.LOGIN]: 'secure',
-};
+type TokenType = Record<AccountTokenReason, {
+	generate: () => string;
+	expiration: number;
+	limit: number;
+}>;
+
+const TOKEN_TYPES = {
+	[AccountTokenReason.ACTIVATION]: {
+		generate: GenerateSimpleToken,
+		expiration: ACTIVATION_TOKEN_EXPIRATION,
+		limit: 1,
+	},
+	[AccountTokenReason.PASSWORD_RESET]: {
+		generate: GenerateSimpleToken,
+		expiration: PASSWORD_RESET_TOKEN_EXPIRATION,
+		limit: 1,
+	},
+	[AccountTokenReason.LOGIN]: {
+		generate: () => nanoid(32),
+		expiration: LOGIN_TOKEN_EXPIRATION,
+		limit: 5,
+	},
+} as const satisfies TokenType;
+
+export interface AccountTokenOwner {
+	onAccountTokenDestroyed(token: AccountToken): void;
+}
 
-const TOKEN_EXPIRATION: Record<AccountTokenReason, number> = {
-	[AccountTokenReason.ACTIVATION]: ACTIVATION_TOKEN_EXPIRATION,
-	[AccountTokenReason.PASSWORD_RESET]: PASSWORD_RESET_TOKEN_EXPIRATION,
-	[AccountTokenReason.LOGIN]: LOGIN_TOKEN_EXPIRATION,
-};
+export class AccountToken implements Readonly<DatabaseAccountToken> {
+	readonly #bound = new Set<AccountTokenOwner>();
+	readonly #reason: AccountTokenReason;
+	#value: string;
+	#expires: number;
 
-const TOKEN_LIMITS: Record<AccountTokenReason, number> = {
-	[AccountTokenReason.ACTIVATION]: 1,
-	[AccountTokenReason.PASSWORD_RESET]: 1,
-	[AccountTokenReason.LOGIN]: 5,
-};
+	#destroyed = false;
+
+	public get value(): string {
+		return this.#value;
+	}
+	public get expires(): number {
+		return this.#expires;
+	}
+	public get reason(): AccountTokenReason {
+		return this.#reason;
+	}
+
+	constructor(token: Readonly<DatabaseAccountToken>) {
+		this.#value = token.value;
+		this.#expires = token.expires;
+		this.#reason = token.reason;
+	}
+
+	public static create(reason: AccountTokenReason): AccountToken {
+		const { generate, expiration } = TOKEN_TYPES[reason];
+		return new AccountToken({
+			value: generate(),
+			expires: Date.now() + expiration,
+			reason,
+		});
+	}
+
+	public getId(): string {
+		return this.value.substring(0, ACCOUNT_TOKEN_ID_LENGTH);
+	}
+
+	public bind(owner: AccountTokenOwner): void {
+		this.#bound.add(owner);
+	}
+
+	public validate(): boolean {
+		if (this.#destroyed)
+			return false;
+
+		if (this.expires <= Date.now()) {
+			this.onDestroy();
+			return false;
+		}
+
+		return true;
+	}
+
+	public onDestroy(): void {
+		if (this.#destroyed)
+			return;
+
+		this.#destroyed = true;
+
+		for (const owner of this.#bound) {
+			owner.onAccountTokenDestroyed(this);
+		}
+		this.#bound.clear();
+	}
+
+	public extend(): boolean {
+		if (this.#destroyed)
+			return false;
+
+		this.#expires = Date.now() + TOKEN_TYPES[this.reason].expiration;
+
+		if (this.value.length > ACCOUNT_TOKEN_ID_LENGTH)
+			this.#value = this.getId() + nanoid(32 - ACCOUNT_TOKEN_ID_LENGTH);
```

Comment:
Why does this happen? It in some weird edge-case can regenerate _part_ of the token, which effectively means all existing tokens become invalid, because they differ from this one. That will however not be easily visible, because they will still share the first 32 characters - the visible id. So the token is seemingly valid on both sides, but invalid as on server the rest of the token is different.
That is definitely not a behavior I would expect from `extend` on an existing token.